couchdb/source/pkg/adapter/adapter.go

```diff
@@ -18,68 +18,87 @@ package adapter
-	EventSource string
-	CouchDbURL  string
-	Username    string
-	Password    string
-	Database    string
-	Namespace   string
-}
+type envConfig struct {
+	adapter.EnvConfig
 
-// Adapter converts incoming CouchDb events to CloudEvents
-type Adapter interface {
-	Start(stopCh <-chan struct{}) error
+	CouchDbCredentialsPath string `envconfig:"COUCHDB_CREDENTIALS" required:"true"`
+	Database               string `envconfig:"COUCHDB_DATABASE" required:"true"`
+	EventSource            string `envconfig:"EVENT_SOURCE" required:"true"`
 }
 
-type adapter struct {
+type couchDbAdapter struct {
+	namespace string
+	ce        cloudevents.Client
+	reporter  source.StatsReporter
+	logger    *zap.SugaredLogger
+
 	source     string
-	ce         cloudevents.Client
-	namespace  string
-	logger     *zap.SugaredLogger
 	couchDB    couchdb.Connection
 	database   string
 	changeArgs couchdb.ChangeArguments
 }
 
-// New creates an adapter to convert incoming CouchDb changes events to CloudEvents and
-// then sends them to the specified Sink
-func New(ceClient cloudevents.Client, logger *zap.SugaredLogger, options *Options) (Adapter, error) {
-	connection, err := couchdb.Connect(options.CouchDbURL, options.Username, options.Password, time.Minute)
+func NewEnvConfig() adapter.EnvConfigAccessor {
+	return &envConfig{}
+}
+
+func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client, reporter source.StatsReporter) adapter.Adapter {
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	rawurl, err := ioutil.ReadFile(env.CouchDbCredentialsPath + "/url")
 	if err != nil {
-		return nil, err
+		logger.Fatal("Missing url key in secret", zap.Error(err))
+	}
+
+	rawusername, err := ioutil.ReadFile(env.CouchDbCredentialsPath + "/username")
+	if err != nil {
+		logger.Fatal("Missing username key in secret", zap.Error(err))
 	}
 
-	a := &adapter{
-		ce:         ceClient,
-		logger:     logger,
-		namespace:  options.Namespace,
+	rawpassword, err := ioutil.ReadFile(env.CouchDbCredentialsPath + "/password")
+	if err != nil {
+		logger.Fatal("Missing username key in secret", zap.Error(err))
+	}
+
+	connection, err := couchdb.Connect(string(rawurl), string(rawusername), string(rawpassword), time.Minute)
+	if err != nil {
+		logger.Fatal("Error creating connection to couchDb", zap.Error(err))
+	}
+
+	return &couchDbAdapter{
+		namespace: env.Namespace,
+		ce:        ceClient,
+		reporter:  reporter,
+		logger:    logger,
+
 		couchDB:    connection,
-		source:     options.EventSource,
-		database:   options.Database,
+		source:     env.EventSource,
+		database:   env.Database,
 		changeArgs: couchdb.ChangeArguments{},
 	}
-
-	return a, nil
 }
 
-func (a *adapter) Start(stopCh <-chan struct{}) error {
+func (a *couchDbAdapter) Start(stopCh <-chan struct{}) error {
 	// Just Poll for now. Other modes (long running, continues) will be provided later on.
 	return wait.PollUntil(2*time.Second, a.send, stopCh)
```

Comment:
why is that `PollUntil` ?

Reply:
Our PromQL does similar - I wonder if that generally can be improved   ( not saying lije CronJob .. .but ... :-) )